### PR TITLE
WIP:  Add initial abstraction layer

### PR DESCRIPTION
**Pull request by krrutkow**
should be able to do things like this now (but I can't test it):
```jl
 	plc = PLC(
		cpu = "compactlogix",
		protocol = "ab-eip",
		gateway = "192.168.1.1",
		path = "1,0",
	)

        myInt = Int32(plc, "myInt")
        myVector = Vector{Int32}(plc, "myVector", 42)
        myString = String(plc, "myString")

        ref = PLCRef{Int32}(plc, "var")
        @info ref[]
        ref[] = 23
```

### CODE REVIEW

**Comment by markhalonen** [comment]
I think the purpose of libplctag splitting `plc_tag_read` and `plc_tag_get_int32` into 2 separate calls is to allow for batching. So if someone did

```
plc_tag_read(tag1)
plc_tag_read(tag2)
plc_tag_get_int32(tag1)
plc_tag_get_int32(tag2)
```
That would result in one communication to the PLC. Does this abstraction keep that feature?

**Comment by markhalonen** [comment]
Ah maybe I'm wrong https://github.com/kyle-github/libplctag/issues/107

**Comment by markhalonen** [comment]
This is also good reading https://github.com/kyle-github/libplctag/wiki/API

File: src/PLCTag.jl [hunk 1]
```diff
@@ -1,81 +1,97 @@
...
+	)), '&')
+	
+	# TODO: create Bool versions as well
+	for T in (
+		:Int8, :UInt8, :Int16, :UInt16,
+		:Int32, :UInt32, :Int64, :UInt64,
+		:Float32, :Float64,
+	)
+		@eval plcget(::Type{$(T)}, args...; kwargs...) = C.$(Symbol("plc_tag_get_", lowercase(String(T))))(args...; kwargs...)
+		@eval plcset(::Type{$(T)}, args...; kwargs...) = C.$(Symbol("plc_tag_set_", lowercase(String(T))))(args...; kwargs...)
```
Comment (markhalonen) [inline src/PLCTag.jl]: Can you explain what this is doing? Is it defining a function `plcget` and `plcset` for each `T`?

Comment (krrutkow) [reply src/PLCTag.jl]: Correct!

**Comment by krrutkow** [comment]
Yeah, the way I interpreted their documentation, the read is what actually gets the data from the PLC, then the get is kind of like a cast on the data region to obtain the desired type.  Obviously all the annoying details of locks and such are dealt with too, but that is the basic concept.

File: src/PLCTag.jl [hunk 2]
```diff
@@ -1,81 +1,97 @@
...
+	Base.length(ref::PLCRef) = ref.count
+	Base.size(ref::PLCRef) = (ref.count,)
+	Base.eltype(ref::PLCRef{T}) where {T} = T
+	Base.iterate(ref::PLCRef, state = 1) = state > length(ref) ? nothing : (ref[state], state+1)
+	
+	function Base.getindex(ref::PLCRef, ind::Int = 1)
+		1 <= ind <= length(ref) || error("Index $(ind) is out of range for PLCRef of count $(ref.count)")
+		
+		code = C.plc_tag_read(ref.tag[], DATA_TIMEOUT)
+		code == C.PLCTAG_STATUS_OK || error("ERROR: Unable to read the data! Got error code $(code): $(unsafe_string(C.plc_tag_decode_error(code)))")
```
Comment (markhalonen) [inline src/PLCTag.jl]: I don't think we want to do this. I think we'll want the async functionality of libplctag, which is invoked by setting `DATA_TIMEOUT` to `0` on `plc_tag_read`. Here is his example of how it's done: https://github.com/kyle-github/libplctag/blob/master/src/examples/async.c

I wrote up a test to see the performance difference on the PLC (I'm at Erico rn):
```julia
using PLCTag

const TAG_PATH = "protocol=ab_eip&gateway=192.168.1.1&path=1,0&cpu=compactlogix&elem_size=4&elem_count=1&name=topTerm&debug=1"
const DATA_TIMEOUT = 5000

const FLOAT_TAG_PATH     = "protocol=ab_eip&gateway=192.168.1.1&path=1,0&cpu=compactlogix&elem_size=4&elem_count=1&name=floater&debug=1"
const FLOAT_DATA_TIMEOUT = 5000

tag = PLCTag.C.plc_tag_create(TAG_PATH, DATA_TIMEOUT)
tag <= 0 && error("ERROR $(PLCTag.C.plc_tag_decode_error(tag)): Could not create tag!")
code = PLCTag.C.plc_tag_status(tag)
code == PLCTag.C.PLCTAG_STATUS_OK || error("ERROR: Error setting up tag internal state. Got error code $(code): $(unsafe_string(PLCTag.C.plc_tag_decode_error(code)))")

float_tag = PLCTag.C.plc_tag_create(FLOAT_TAG_PATH, DATA_TIMEOUT)
float_tag <= 0 && error("ERROR $(PLCTag.C.plc_tag_decode_error(float_tag)): Could not create tag!")
code = PLCTag.C.plc_tag_status(float_tag)
code == PLCTag.C.PLCTAG_STATUS_OK || error("ERROR: Error setting up tag internal state. Got error code $(code): $(unsafe_string(PLCTag.C.plc_tag_decode_error(code)))")



function read_sync()
	@time begin
	# Do 10 reads
		for i in 1:100
			code = PLCTag.C.plc_tag_read(tag, 5000)
			code == PLCTag.C.PLCTAG_STATUS_OK || error("ERROR: Unable to read the data! Got error code $(code): $(unsafe_string(PLCTag.C.plc_tag_decode_error(code)))")
			float_code = PLCTag.C.plc_tag_read(float_tag, 5000)
			float_code == PLCTag.C.PLCTAG_STATUS_OK || error("ERROR: Unable to read the data! Got error code $(float_code): $(unsafe_string(PLCTag.C.plc_tag_decode_error(float_code)))")


			val = PLCTag.C.plc_tag_get_int32(tag, 0)
			float_val = PLCTag.C.plc_tag_get_float32(float_tag, 0)
			val == 1 || error("Did not get 1")
			isapprox(float_val, 1.123) || error("Did not get 1.123")
		end
	end
end

function read_async()
	@time begin
		# Do 10 reads
		for i in 1:100
			code = PLCTag.C.plc_tag_read(tag, 0)
			float_code = PLCTag.C.plc_tag_read(float_tag, 0)

			code == PLCTag.C.PLCTAG_STATUS_OK || code == PLCTag.C.PLCTAG_STATUS_PENDING || error("ERROR: Unable to read the data! Got error code $(code): $(unsafe_string(PLCTag.C.plc_tag_decode_error(code)))")
			
			float_code == PLCTag.C.PLCTAG_STATUS_OK || code == PLCTag.C.PLCTAG_STATUS_PENDING || error("ERROR: Unable to read the data! Got error code $(float_code): $(unsafe_string(PLCTag.C.plc_tag_decode_error(float_code)))")
	

			while true
				if PLCTag.C.plc_tag_status(tag) == PLCTag.C.PLCTAG_STATUS_OK && PLCTag.C.plc_tag_status(float_tag) == PLCTag.C.PLCTAG_STATUS_OK
					break
				end
				sleep(.001)
			end

	
			val = PLCTag.C.plc_tag_get_int32(tag, 0)
			float_val = PLCTag.C.plc_tag_get_float32(float_tag, 0)
			val == 1 || error("Did not get 1")
			isapprox(float_val, 1.123) || error("Did not get 1.123")
		end
	end
end
```

This is only using 2 tags.

As you can see, using async is twice as fast:
```
julia> read_async()
  0.493761 seconds (5.45 k allocations: 123.063 KiB)

julia> read_sync()
  0.891796 seconds (4.30 k allocations: 92.188 KiB)

julia> read_async()
  0.451472 seconds (5.38 k allocations: 121.359 KiB)

julia> read_sync()
  0.852679 seconds (4.30 k allocations: 92.188 KiB)

julia> read_async()
  0.479070 seconds (5.44 k allocations: 122.891 KiB)

julia> read_sync()
  0.959426 seconds (4.30 k allocations: 92.188 KiB)

julia> read_async()
  0.442862 seconds (5.33 k allocations: 120.656 KiB)

julia> read_sync()
  0.847073 seconds (4.30 k allocations: 92.188 KiB)

julia> read_async()
  0.484797 seconds (5.36 k allocations: 121.016 KiB)

julia>
```

I'm guessing the performance will be even more dramatic for many tags. For example, 10 tags might be the same timing for async as 2 tags if network is driving performance.

So I think this abstraction is oversimplifying their api and we'll want async functionality.

Comment (krrutkow) [reply src/PLCTag.jl]: Yeah, that is good information, we can probably switch to using Futures for getting/setting values asynchronously then.  It probably has something to do with their internal threads and locking in libplctag.

Comment (krrutkow) [reply src/PLCTag.jl]: Yikes, please test how async scales with number of tags.  Its starting at 0.5 seconds to query 100 times, hopefully doing 10 tags is still 0.5 seconds and not 5x more.

Comment (markhalonen) [reply src/PLCTag.jl]: yeah, we should have tests for all data types and also performance tests for 1000 tags or something. They are setting up a computer with RSLogix so I'll be able to set up a test tag project to load onto the PLC.

Comment (markhalonen) [reply src/PLCTag.jl]: I also did these tests over a WiFi network, that probably slowed things down

**Comment by krrutkow** [comment]
@markhalonen I made up a benchmarking script (not tested of course) to test 4 usages at the moment.  SoA (arrayed tags) and AoS (individual tags), and synchronous versus asynchronous reading.  Eventually we will want to add writing tests too.

|        | sync | async |
| ----- |:------:|:------:|
| SoA | :heavy_check_mark: | :heavy_check_mark: |
| AoS | :heavy_check_mark: | :heavy_check_mark: |

**Comment by krrutkow** [comment]
Sorry, if you didn't know the terms, AoS is array-of-structures and SoA is structure-of-arrays and look like this in code:
```jl
struct AoS
   field1::Int32
   field2::Float32
end
aos = [
   AoS(1, 2.3),
   AoS(4, 5.6),
]

struct SoA
   field1::Vector{Int32}
   field2::Vector{Float32}
end
soa = SoA(
   Int32[1, 4],
   Float32[2.3, 5.6],
)
```

Also, using the SoA paradigm means that literally every instance of a field is pulled from the PLC even if only 1 of them is actually needed, while AoS allows for the individual fields to be pulled so reading many instances requires more overhead.

**Comment by markhalonen** [comment]
After my changes to make stuff work, here are the results:
```
julia> TEST.timeAll()
[ Info: synchronously reading individual tags
 19.475119 seconds (105.00 k allocations: 2.213 MiB)
[ Info: asynchronously reading individual tags
  0.678183 seconds (92.40 k allocations: 2.038 MiB)
[ Info: synchronously reading arrayed tags
  1.934830 seconds (51.00 k allocations: 1.114 MiB)
[ Info: asynchronously reading arrayed tags
  0.470877 seconds (51.20 k allocations: 1.129 MiB)

julia>
```

**Comment by markhalonen** [comment]
Here is the exported tag csv I used:
[benchmark_stuff.zip](https://github.com/laurium-labs/PLCTag.jl/files/3691633/benchmark_stuff.zip)

This is the RSLogix CSV format. So we can import tags that are in this format (that's what I did to automate tag creation).

**Comment by krrutkow** [comment]
Awesome info!  If you can, try scaling the #fields or the #instances to see how each of those affect performance.

**Comment by markhalonen** [comment]
Ok won't have time for that today, headed out to the hunting camp. Should be easy enough to do though

**Comment by markhalonen** [comment]
```
julia> module TEST ; include("bin/benchmark.jl") ; end
Main.TEST

julia> TEST.timeAll(fields=1, instances=1)
[ Info: Running tests with fields=1 and instances=1 and iterations=100
[ Info: synchronously reading individual tags
  0.375241 seconds (2.10 k allocations: 45.313 KiB)
[ Info: asynchronously reading individual tags
  0.452403 seconds (3.45 k allocations: 78.484 KiB)
[ Info: synchronously reading arrayed tags
  0.407366 seconds (2.10 k allocations: 45.313 KiB)
[ Info: asynchronously reading arrayed tags
  0.443861 seconds (3.45 k allocations: 78.484 KiB)

julia> TEST.timeAll(fields=1, instances=5)
[ Info: Running tests with fields=1 and instances=5 and iterations=100
[ Info: synchronously reading individual tags
  1.896455 seconds (10.50 k allocations: 226.563 KiB)
[ Info: asynchronously reading individual tags
  0.446915 seconds (10.66 k allocations: 241.656 KiB)
[ Info: synchronously reading arrayed tags
  0.387040 seconds (5.70 k allocations: 126.563 KiB)
[ Info: asynchronously reading arrayed tags
  0.443668 seconds (7.02 k allocations: 159.219 KiB)

julia> TEST.timeAll(fields=5, instances=5)
[ Info: Running tests with fields=5 and instances=5 and iterations=100
[ Info: synchronously reading individual tags
 10.017503 seconds (52.50 k allocations: 1.106 MiB)
[ Info: asynchronously reading individual tags
  0.479793 seconds (46.74 k allocations: 1.031 MiB)
[ Info: synchronously reading arrayed tags
  1.928708 seconds (28.50 k allocations: 632.813 KiB)
[ Info: asynchronously reading arrayed tags
  0.466863 seconds (28.64 k allocations: 647.063 KiB)

julia> TEST.timeAll(fields=10, instances=5)
[ Info: Running tests with fields=10 and instances=5 and iterations=100
[ Info: synchronously reading individual tags
 20.122860 seconds (105.00 k allocations: 2.213 MiB, 0.07% gc time)
[ Info: asynchronously reading individual tags
  0.726966 seconds (92.45 k allocations: 2.039 MiB)
[ Info: synchronously reading arrayed tags
  4.058523 seconds (57.00 k allocations: 1.236 MiB)
[ Info: asynchronously reading arrayed tags
  0.499603 seconds (55.79 k allocations: 1.230 MiB)

julia> TEST.timeAll(fields=5, instances=10)
[ Info: Running tests with fields=5 and instances=10 and iterations=100
[ Info: synchronously reading individual tags
 18.746826 seconds (105.00 k allocations: 2.213 MiB)
[ Info: asynchronously reading individual tags
  0.683087 seconds (92.37 k allocations: 2.037 MiB)
[ Info: synchronously reading arrayed tags
  1.928735 seconds (51.00 k allocations: 1.114 MiB, 0.35% gc time)
[ Info: asynchronously reading arrayed tags
  0.458050 seconds (51.12 k allocations: 1.127 MiB)

julia> TEST.timeAll(fields=10, instances=10)
[ Info: Running tests with fields=10 and instances=10 and iterations=100
[ Info: synchronously reading individual tags
 40.313049 seconds (210.00 k allocations: 4.425 MiB)
[ Info: asynchronously reading individual tags
  0.906204 seconds (187.84 k allocations: 4.108 MiB)
[ Info: synchronously reading arrayed tags
  4.042196 seconds (146.00 k allocations: 2.899 MiB)
[ Info: asynchronously reading arrayed tags
  0.504560 seconds (146.89 k allocations: 2.925 MiB)
```

**Comment by bhalonen** [comment]
Instances mean number of tags, fields mean how big the "struct" is?

**Comment by markhalonen** [comment]
`instances` is number of elements in the array (in array mode). So X number of fields, each with Y instances for arrayed.

For individual, just X * Y number of tags.

**Comment by markhalonen** [comment]
This is what I think we're going to be reading, about 300 individual tags in async mode.
```
julia> TEST.timeAll(fields=300, instances=1)
[ Info: Running tests with fields=300 and instances=1 and iterations=100
[ Info: asynchronously reading individual tags
  2.147382 seconds (847.97 k allocations: 16.653 MiB, 0.29% gc time)

julia>
```
So this has 100 iterations, meaning we read the 300 tags 100 times in 2.14 seconds, or about 50 times per second.

This is with all `DINT` types and nothing else running on the PLC I don't think.

**Comment by bhalonen** [comment]
Should be golden for collecting once every second or 2 then.

**Comment by krrutkow** [comment]
Added `bin/hacking.jl` to demonstrate the `PLCBinding` usage.  Nothing is set in stone, but it provides a function call abstraction binding with the PLC to do some testing with.  One concern is that the benchmarking only focused on the performance of bulk reads, not mixtures of reads and writes.  So using the PLCBinding mechanism for "calling" the PLC provides opportunity for doing some new performance testing.  Also, beware, the code is untested, just wanted to get it pushed.

File: src/PLCTag.jl [hunk 3]
```diff
@@ -72,26 +78,93 @@ module PLCTag
...
 	for T in (
+		:Bool,
 		:Int8, :UInt8, :Int16, :UInt16,
 		:Int32, :UInt32, :Int64, :UInt64,
 		:Float32, :Float64,
 	)
-		@eval plcget(::Type{$(T)}, args...; kwargs...) = C.$(Symbol("plc_tag_get_", lowercase(String(T))))(args...; kwargs...)
-		@eval plcset(::Type{$(T)}, args...; kwargs...) = C.$(Symbol("plc_tag_set_", lowercase(String(T))))(args...; kwargs...)
+		if T === Bool
+			# TODO: determine what "bool" size is on a PLC
```
Comment (markhalonen) [inline src/PLCTag.jl]: The size of a bool on the PLC varies for the following cases:
- independent tag
- part of an array
- part of a UDT (user defined type)

https://github.com/kyle-github/libplctag/issues/82

For this reason, libplctag does not have support for bools. Our solution is to ban the usage of bools on the PLC, and only use SINT (1 byte).

**Comment by markhalonen** [comment]
This thought just occurred to me.. should we be implementing the `Return` value logic in this library? It seems specific to our use case. I think we plan to submit this as the Julia wrapper to libplctag.

**Pull request merged**